### PR TITLE
BridgePolicy - Deal With SpaceInfo Instead of Space Names

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch changes `BridgePolicy` so that we begin to pass around `SpaceInfo` instead of space name strings. It is not the final form of this logic - completion depends upon finalisation of subnet space designation being by ID instead of name.

A space cache is created upon `BridgePolicy` instantiation, which is used in this intermediate form to look up spaces by space names.

## QA steps

Bridging via Constraints:

- Bootstrap guimaas.
- `juju add-machine --constraints spaces=default,space-alt2`.
- `juju add-machine lxd:0 --constraints spaces=default,space-alt2`.
- Once the container has started, `juju ssh` to the host machine and check that bridges have been created for the NIC in each space.

Bridging via Endpoints:

- Bootstrap guimaas (or destroy previous model and create a new one).
- `juju add-machine --constraints spaces=default,space-alt2`.
- `juju deploy keystone --bind "public=default internal=space-alt2 admin=default shared-db=space-alt2" --to lxd:0`
- Once the container has started, `juju ssh` to the host machine and check that bridges have been created for the NIC in each space.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 1]
```diff
@@ -247,11 +247,11 @@ func NewProvisionerAPIV9(st *state.State, resources facade.Resources, authorizer
 	return &ProvisionerAPIV9{provisionerAPI}, nil
 }
 
-func (p *ProvisionerAPI) getMachine(canAccess common.AuthFunc, tag names.MachineTag) (*state.Machine, error) {
+func (api *ProvisionerAPI) getMachine(canAccess common.AuthFunc, tag names.MachineTag) (*state.Machine, error) {
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioner.go]: Making this receive consistent adds noise for which the author apologises.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 2]
```diff
@@ -929,6 +929,19 @@ type Machine interface {
 	MachineTag() names.MachineTag
 }
 
+// BridgePolicy is an indirection for containerizer.BridgePolicy.
+type BridgePolicy interface {
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioner.go]: Before introducing this indirection, we were mocking expectations for arguments passed to `BridgePolicy` methods. This is obviously leaky as hell - we shouldn't be mocking behaviour expected by a dependency of ours, rather just mocking the dependency.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 3]
```diff
@@ -954,20 +968,25 @@ type perContainerHandler interface {
...
 	if errors.IsNotProvisioned(err) {
-		err = errors.NotProvisionedf("cannot prepare container %s config: host machine %q", handler.ConfigType(), hostMachine)
-		return err
+		return errors.NotProvisionedf("cannot prepare container %s config: host machine %q",
+			handler.ConfigType(), hostMachine)
 	} else if err != nil {
 		return errors.Trace(err)
 	}
 
+	policy, err := containerizer.NewBridgePolicy(env, api.st)
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioner.go]: This is now created once per batch of containers instead of being created each time inside `ProcessOneContainer`.

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 4]
```diff
@@ -1142,24 +1110,85 @@ func (ctx *prepareOrGetContext) ProcessOneContainer(
 	return nil
 }
 
-func (p *ProvisionerAPI) prepareOrGetContainerInterfaceInfo(args params.Entities, maintain bool) (params.MachineNetworkConfigResults, error) {
+// infoForDevice returns interface information for a link-layer device.
+func (ctx *prepareOrGetContext) infoForDevice(
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioner.go]: Just an extraction to cut down the size of `ProcessOneContainer`.

File: apiserver/facades/agent/provisioner/provisioner_test.go [hunk 5]
```diff
@@ -1421,28 +1421,90 @@ func (s *withoutControllerSuite) TestContainerManagerConfigDefaults(c *gc.C) {
 	})
 }
 
-type withImageMetadataSuite struct {
-	provisionerSuite
-}
+func (s *withoutControllerSuite) TestWatchMachineErrorRetry(c *gc.C) {
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioner_test.go]: Some non-functional churn here due to method relocation, but having tests from different suite spread around was annoying.

File: apiserver/facades/agent/provisioner/shim.go [hunk 6]
```diff
@@ -1,42 +0,0 @@
-// Copyright 2018 Canonical Ltd.
-// Licensed under the AGPLv3, see LICENCE file for details.
-
-package provisioner
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/shim.go]: This shims and their mocks were unused, and so were deleted.

File: network/containerizer/bridgepolicy_integration_test.go [hunk 7]
```diff
@@ -76,9 +74,7 @@ func (s *bridgePolicyStateSuite) SetUpTest(c *gc.C) {
 	var err error
 	m, err := s.State.AddMachine("quantal", state.JobHostUnits)
 	c.Assert(err, jc.ErrorIsNil)
-	s.machine = &containerizer.MachineShim{Machine: m}
-
-	s.bridgePolicy = containerizer.NewBridgePolicy(cfg(c, 13, "provider"))
```
Comment (manadart) [inline network/containerizer/bridgepolicy_integration_test.go]: This was moved off the suite and created for each test, because we have to do it _after_ any spaces are created so they end up in the cache.

File: apiserver/facades/agent/provisioner/provisioninginfo.go [hunk 8]
```diff
@@ -29,15 +29,15 @@ import (
 )
 
 // ProvisioningInfo returns the provisioning information for each given machine entity.
-func (p *ProvisionerAPI) ProvisioningInfo(args params.Entities) (params.ProvisioningInfoResults, error) {
+func (api *ProvisionerAPI) ProvisioningInfo(args params.Entities) (params.ProvisioningInfoResults, error) {
```
Comment (manadart) [inline apiserver/facades/agent/provisioner/provisioninginfo.go]: All changes to this file were receiver renaming is can be ignored by reviewer(s).

File: network/containerizer/bridgepolicy.go [hunk 9]
```diff
@@ -41,28 +50,35 @@ type BridgePolicy struct {
 
 // NewBridgePolicy returns a new BridgePolicy for the input environ config
 // getter and state indirection.
-func NewBridgePolicy(cfgGetter environs.ConfigGetter) *BridgePolicy {
+func NewBridgePolicy(cfgGetter environs.ConfigGetter, st SpaceBacking) (*BridgePolicy, error) {
```
Comment (achilleasa) [inline network/containerizer/bridgepolicy.go]: We should probably reach some team consensus about naming these interfaces in a consistent manner. I have seen references in the code where we call these Backends, Contexts, APIs etc.

File: network/containerizer/bridgepolicy.go [hunk 10]
```diff
@@ -80,15 +96,22 @@ func (p *BridgePolicy) FindMissingBridgesForContainer(
...
+	fanNotFound := guestSpaceSet.Difference(fanSpacesFound)
+
 	if p.containerNetworkingMethod == "fan" {
 		if fanNotFound.IsEmpty() {
 			// Nothing to do, just return success
 			return nil, 0, nil
 		} else {
 			return nil, 0, errors.Errorf("host machine %q has no available FAN devices in space(s) %s",
-				m.Id(), network.QuoteSpaceSet(fanNotFound))
+				host.Id(), network.QuoteSpaceSet(fanNotFound))
```
Comment (achilleasa) [inline network/containerizer/bridgepolicy.go]: Drive-by fix opportunity: the else block is redundant and can be replaced with the return statement. Also, you can convert the following nested else block into an `else if` block

File: network/containerizer/bridgepolicy.go [hunk 11]
```diff
@@ -380,11 +405,13 @@ func (p *BridgePolicy) PopulateContainerLinkLayerDevices(m Machine, containerMac
 			}
 		}
 	}
-	missingSpace := containerSpaces.Difference(spacesFound)
+
+	guestSpaceSet := set.NewStrings(guestSpaces.Names()...)
+	missingSpace := guestSpaceSet.Difference(spacesFound)
```
Comment (achilleasa) [inline network/containerizer/bridgepolicy.go]: missingSpace**s**?

File: network/containerizer/bridgepolicy.go [hunk 12]
```diff
@@ -380,11 +405,13 @@ func (p *BridgePolicy) PopulateContainerLinkLayerDevices(m Machine, containerMac
...
 
 	// Check if we are missing "" and can fill it in with a local bridge
 	if len(missingSpace) == 1 && missingSpace.Contains("") && p.containerNetworkingMethod == "local" {
```
Comment (achilleasa) [inline network/containerizer/bridgepolicy.go]: Perhaps the `""` could be replaced by `network.DefaultSpaceName` to be more consistent?

File: network/containerizer/bridgepolicy.go (hunk 11 above)
Comment (manadart) [reply network/containerizer/bridgepolicy.go]: Done.

File: network/containerizer/bridgepolicy.go (hunk 10 above)
Comment (manadart) [reply network/containerizer/bridgepolicy.go]: Done.

File: network/containerizer/bridgepolicy.go (hunk 12 above)
Comment (manadart) [reply network/containerizer/bridgepolicy.go]: Done.

**Comment by manadart** [comment]
Tests in `state/linklayerdevices_test.go` are green now.

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$
